yt/utilities/lib/image_utilities.pyx

```diff
@@ -36,6 +41,370 @@ def add_points_to_greyscale_image(
+    jmin = int(floor(lx * Nx))
+    if jmin < 0: jmin = 0
+    jmax = int(ceil(rx * Nx))
+    if jmax >= Nx: jmax = Nx - 1
+
+    kmin = int(floor(ly * Ny))
+    if kmin < 0: kmin = 0
+    kmax = int(ceil(ry * Ny))
+    if kmax >= Ny: kmax = Ny - 1
+
+    # If the cell is fully within one pixel
+    if (jmax == jmin + 1) and (kmax == kmin + 1):
+        buffer[jmin][kmin] += w * q * dx
+        buffer_weight[jmin][kmin] += w * dx
+        return
+
+    depth = int(ceil(log2(2 * sqrt(3) * dx * float(max(Nx, Ny)))))
+    if depth < 1: depth = 1
+    if depth >= max_depth: depth = max_depth - 1
+
+    jmax = min(Nsx, 1 << depth)
+    kmax = min(Nsy, 1 << depth)
+
+    dx_loc = cell_max_width / jmax
+    dy_loc = cell_max_width / kmax
+
+    # with gil: print(jmax, kmax)
+
+    for j in range(jmax):
+        xx1 = ((j - jmax / 2.) * dx_loc + x) * Nx
+        xx2 = ((j + 1 - jmax / 2.) * dx_loc + x) * Nx
+
+        jj1 = int(xx1)
+        jj2 = int(xx2)
+        # The subcell is out of the projected area
+        if jj2 < 0 or jj1 >= Nx: continue
+
+        dvx = max(0, min(1, (jj2 - xx1) / (xx2 - xx1)))
+
+        for k in range(kmax):
+            if stamp_mask[depth, j, k] == 0:
+                continue
+
+            yy1 = ((k - kmax / 2.) * dy_loc + y) * Ny
+            yy2 = ((k + 1 - kmax / 2.) * dy_loc + y) * Ny
+
+            kk1 = int(yy1)
+            kk2 = int(yy2)
+            # The subcell is out of the projected area
+            if kk2 < 0 or kk1 >= Ny: continue
+
+            # with gil:
+            #     print(f"{j=} {k=} {xx1=} {xx2=} {yy1=} {yy2=} {jj1=} {jj2=} {kk1=} {kk2=}")
+
+            sw =  stamp[depth, j, k] * w * dx
+            swq = sw * q
+
+            dvy = max(0, min(1, (kk2 - yy1) / (yy2 - yy1)))
+
+            if jj1 >= 0 and kk1 >= 0:
+                buffer[jj1][kk1]        += swq * dvx * dvy
+                buffer_weight[jj1][kk1] +=  sw * dvx * dvy
+
+            if jj1 >= 0 and kk2 < Ny:
+                buffer[jj1][kk2]        += swq * dvx * (1 - dvy)
+                buffer_weight[jj1][kk2] +=  sw * dvx * (1 - dvy)
+
+            if jj2 < Nx and kk1 >= 0:
+                buffer[jj2][kk1]        += swq * (1 - dvx) * dvy
+                buffer_weight[jj2][kk1] +=  sw * (1 - dvx) * dvy
+
+            if jj2 < Nx and kk2 < Ny:
+                buffer[jj2][kk2]        += swq * (1 - dvx) * (1 - dvy)
+                buffer_weight[jj2][kk2] +=  sw * (1 - dvx) * (1 - dvy)
+
+@cython.boundscheck(False)
+@cython.cdivision(True)
+@cython.wraparound(False)
+def add_cells_to_image_offaxis(
+    *,
+    const np.float64_t[:, ::1] Xp,
+    const np.float64_t[::1] dXp,
+    const np.float64_t[::1] qty,
+    const np.float64_t[::1] weight,
+    const np.float64_t[:, :] rotation,
+    np.float64_t[:, ::1] buffer,
+    np.float64_t[:, ::1] buffer_weight,
+    const int Nx,
+    const int Ny,
+    const int Npix_min = 4,
+):
+    cdef np.ndarray[np.float64_t, ndim=1] center = np.array([0.5, 0.5, 0.5])
```

Comment:
why is this fixed rather than an argument?

Reply:
In the "stamping" approach, I compute the projection of a cube of size 1 centred on [0.5, 0.5, 0.5]. This is then "stamped" at the location of each cell. The centre here relates to the centre of this reference cube, not to where the centre actually is. This is done from the caller's side of things.